### PR TITLE
Update endtime instead of insert a new line in iplog table

### PR DESCRIPTION
**Pull request by fdurand**
## Description

Update end_time in iplog instead of closing iplog entry and opening a new one.
## Impacts

Less entries in iplog
## Delete branch after merge

YES

### CODE REVIEW

File: lib/pf/api.pm [hunk 1]
```diff
@@ -97,10 +97,26 @@ sub soh_authorize : Public {
 }
 
 sub update_iplog : Public {
-    my ( $class, $srcmac, $srcip, $lease_length ) = @_;
+    my ($class, %postdata) = @_;
+    my @require = qw(mac ip lease_length oldmac oldip);
+    my @found = grep {exists $postdata{$_}} @require;
+    return unless @require == @found;
```
Comment (julsemaan) [inline lib/pf/api.pm]: We shouldn't have silent returns like that. It should at least log a message.

File: NEWS.asciidoc [hunk 2]
```diff
@@ -42,6 +42,7 @@ Enhancements
 * Reduce deadlocks caused by the cleaning of the iplog table
 * Reduce deadlocks caused by the cleaning of the locationlog table
 * Reorganize the portal profile option page
+* Reduce the numbers of entries in iplog tables (update endtime instead of close and insert a new line)
```
Comment (julsemaan) [inline NEWS.asciidoc]: *in the iplog table
*instead of closing and inserting...

File: lib/pf/api.pm [hunk 3]
```diff
@@ -409,12 +424,40 @@ sub node_information : Public {
...
     my @require = qw(mac);
     my @found = grep {exists $postdata{$_}} @require;
-    return unless @require == @found;
+    return unless validate_argv(\@require,  \@found);
 
     my $node_info = pf::node::node_view($postdata{'mac'});
     return $node_info;
 }
 
+=head2 validate_arg
```
Comment (julsemaan) [inline lib/pf/api.pm]: This says validate_arg but it's validate_argv

**Comment by julsemaan** [comment]
You're missing the part in pfdhcplistener

```

diff --git a/sbin/pfdhcplistener b/sbin/pfdhcplistener
index 99e62d9..5d481b6 100755
--- a/sbin/pfdhcplistener
+++ b/sbin/pfdhcplistener
@@ -532,9 +532,9 @@ sub update_iplog {
     }

     my $oldip  = mac2ip($srcmac);
+    my $oldmac = ip2mac($srcip);

     if ( $oldip && $oldip ne $srcip ) {
-        my $oldmac = ip2mac($srcip);
         my $view_mac = node_view($srcmac);
         my $firewallsso = pf::firewallsso->new;
         $firewallsso->do_sso('Stop',$oldmac,$oldip,undef);
@@ -542,7 +542,7 @@ sub update_iplog {
             $apiclient->notify('ipset_node_update',$oldip, $srcip, $srcmac);
         }
     }
-    $apiclient->notify('update_iplog', $srcmac, $srcip, $lease_length );
+    $apiclient->notify('update_iplog', ( mac => $srcmac, ip => $srcip, lease_length => $lease_length, oldmac => $oldmac, oldip => $oldip ) );
 }

 =item process_fingerprint

```

**Comment by julsemaan** [comment]
With the patch above, it works.
It adds new entries and update old ones instead of inserting new ones.

Also tested migrating from old way to new way with existing tables and it stops the iplog explosions.

**Comment by fdurand** [comment]
Ready to merge ?

**Comment by julsemaan** [comment]
Yes,

I'll create the patch for the maintenance

**Pull request merged**